botocore/docs/docstring.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from botocore.docs.method import document_model_driven_method
+from botocore.docs.waiter import document_wait_method
+from botocore.docs.bcdoc.restdoc import DocumentStructure
+
+
+class LazyLoadedDocstring(str):
+    """Used for lazily loading docstrings
+
+    You can instantiate this class and assign it to a __doc__ value.
+    The docstring will not be generated till accessed via __doc__ or
+    help(). Note that all docstring classes **must** subclass from
+    this class. It cannot be used directly as a docstring.
+    """
+    def __init__(self, *args, **kwargs):
+        """
+        The args and kwargs are the same as the underlying document
+        generation function. These just get proxied to the underlying
+        function.
+        """
+        super(LazyLoadedDocstring, self).__init__()
+        self._gen_args = args
+        self._gen_kwargs = kwargs
+        self._docstring = None
+        self._docstring_writer = self._raise_docstring_not_implemented
```

Comment:
I am a little lost here. Comments seem to differ.

@mtdowling I set the ``_raise_docstring_not_implemented`` method to be informative when developing or building subclasses on top of this base class. I could always set ``_docstring_writer`` to ``None`` and that will lead to a ``None`` object is not callable error. I just want the class to fail to generate documentation if the subclass does not specify a ``_docstring_writer`` to help with development and avoiding bugs as there are going to be a fair amount of Docstring class that will subclass from it in the future. What would you suggest?

@jamesls
It sounds like I am doing what you suggest. For example take the [ClientMethodDocstring](https://github.com/boto/botocore/pull/643/files#diff-5989425b25925188aeb4541546e0e61bR84) class. How does it differ from what you are suggesting?